### PR TITLE
fix: add unsupported status condition for filters within BackendRef

### PR DESCRIPTION
**Pull request by cnvergence**
**What type of PR is this?**

- highlight unsupported BackendRef filters using a status condition
- introduce BackendRefContext in translator

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes 
- https://github.com/envoyproxy/gateway/issues/2572

### CODE REVIEW

File: internal/gatewayapi/route.go [hunk 1]
```diff
@@ -173,6 +173,15 @@ func (t *Translator) processHTTPRouteRules(httpRoute *HTTPRouteContext, parentRe
 		dstAddrTypeMap := make(map[ir.DestinationAddressType]int)
 
 		for _, backendRef := range rule.BackendRefs {
+			if len(backendRef.Filters) > 0 {
```
Comment (arkodg) [inline internal/gatewayapi/route.go:176]: should we `continue`, we have been failing fast and early in other places

File: internal/gatewayapi/route.go [hunk 2]
```diff
@@ -173,6 +173,15 @@ func (t *Translator) processHTTPRouteRules(httpRoute *HTTPRouteContext, parentRe
...
+				parentRef.SetCondition(httpRoute,
+					gwapiv1.RouteConditionResolvedRefs,
+					metav1.ConditionFalse,
+					"UnsupportedRefValue",
+					"The filters field within BackendRef is not supported",
+				)
+			}
```
Comment (arkodg) [inline internal/gatewayapi/route.go:183]: probably also need to add this for tcproute, tlsroute and udproute ?

Comment (cnvergence) [reply internal/gatewayapi/route.go:183]: thought so, but they don't have it https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.BackendRef

File: internal/gatewayapi/route.go (hunk 1 above)
Comment (cnvergence) [reply internal/gatewayapi/route.go:176]: we could, let's fail fast then on this

File: internal/gatewayapi/route.go (hunk 2 above)
Comment (arkodg) [reply internal/gatewayapi/route.go:184]: thanks for checking !

File: internal/gatewayapi/route.go (hunk 1 above)
Comment (arkodg) [reply internal/gatewayapi/route.go:176]: rethinking again, this code should live inside [validateBackendRef](https://github.com/envoyproxy/gateway/blob/ac0243749b01d1d60b32bae88cb0effb71b61102/internal/provider/kubernetes/helpers.go#L215) which gets called here
https://github.com/envoyproxy/gateway/blob/ac0243749b01d1d60b32bae88cb0effb71b61102/internal/gatewayapi/route.go#L1080

which automatically converts the `weight` into a `invalidWeight` ensuring that a percentage of traffic gets 503d

Comment (cnvergence) [reply internal/gatewayapi/route.go:176]: looks good, but the field Filters is within gwapiv1.HTTPBackendRef or gwapiv1a1.GRPCBackendRef.
we process BackendRef based on gwapiv1.BackendRef

Comment (arkodg) [reply internal/gatewayapi/route.go:176]: we've had to do something similar for xRoutes, so we created `RouteContext`
https://github.com/envoyproxy/gateway/blob/cc88e9ccd486d14af48138f733ad70ee61a581c6/internal/gatewayapi/contexts.go#L155

may need something similar here

Comment (cnvergence) [reply internal/gatewayapi/route.go:176]: if we want to extend it, then sure
thanks for explaining

File: internal/gatewayapi/contexts.go [hunk 3]
```diff
@@ -417,3 +417,27 @@ func (r *RouteParentContext) HasCondition(route RouteContext, condType gwapiv1.R
 	}
 	return false
 }
+
+// BackendRefContext represents a generic BackendRef object that can reference Gateway objects.
```
Comment (arkodg) [inline internal/gatewayapi/contexts.go:421]: id suggest this be an interface with 2 methods
* GetBackendRef()
* GetFilters()

and the implementation could look similar to https://github.com/envoyproxy/gateway/blob/c5cd964da2f2ec1951ba72d45b54eef8246e96ee/internal/gatewayapi/contexts.go#L216
```


e.g. reflect.ValueOf(route).Elem().FieldByName("BackendRef")

Comment (cnvergence) [reply internal/gatewayapi/contexts.go:421]: I think the struct would be still needed or am I wrong?

**Review by arkodg — APPROVED** [review]
LGTM, thanks for incorporating the feedback

**Comment by cnvergence** [comment]
sure, found the context in https://github.com/envoyproxy/gateway/issues/1252

**Review by arkodg — APPROVED** [review]

**Review by shawnh2 — APPROVED** [review]

**Pull request merged**